File: scripts/count_lines.py
#!/usr/bin/env python3
"""统计项目代码行数：源代码 vs 注释 vs 空白"""

import os
import sys
import re

ROOT = "/Users/linlay/Project/zenmind/agent-platform"

EXCLUDE_DIRS = {".git", ".zenmind", "vendor", "node_modules", "target", "__pycache__"}

def count_go(filepath):
    """统计 Go 文件的代码行、注释行、空白行"""
    with open(filepath, "r", encoding="utf-8", errors="replace") as f:
        lines = f.readlines()

    code, comment, blank = 0, 0, 0
    in_block = False

    for line in lines:
        stripped = line.strip()
        if stripped == "":
            blank += 1
            continue

        if in_block:
            comment += 1
            if "*/" in stripped:
                in_block = False
            continue

        # 检查块注释开始
        if stripped.startswith("/*") or stripped.startswith("//*"):
            comment += 1
            if "*/" not in stripped:
                in_block = True
            continue

        if stripped.startswith("//"):
            comment += 1
            continue

        # 检查行内 /* */
        if "/*" in stripped:
            comment += 1
            if "*/" not in stripped:
                in_block = True
            continue

        code += 1

    return code, comment, blank

def count_rust(filepath):
    """统计 Rust 文件的代码行、注释行、空白行（同 Go）"""
    return count_go(filepath)

def count_yaml(filepath):
    """统计 YAML 文件的代码行、注释行、空白行"""
    with open(filepath, "r", encoding="utf-8", errors="replace") as f:
        lines = f.readlines()

    code, comment, blank = 0, 0, 0

    for line in lines:
        stripped = line.strip()
        if stripped == "":
            blank += 1
        elif stripped.startswith("#"):
            comment += 1
        else:
            code += 1

    return code, comment, blank

def count_markdown(filepath):
    """统计 Markdown（全算文档/注释，不算代码）"""
    with open(filepath, "r", encoding="utf-8", errors="replace") as f:
        lines = f.readlines()

    code, comment, blank = 0, 0, 0

    for line in lines:
        stripped = line.strip()
        if stripped == "":
            blank += 1
        else:
            comment += 1  # Markdown 全部算文档

    return code, comment, blank

def main():
    counters = {"go": count_go, "rs": count_rust, "yml": count_yaml, "yaml": count_yaml, "md": count_markdown}

    totals = {"code": 0, "comment": 0, "blank": 0, "files": 0}
    by_ext = {}

    for root, dirs, files in os.walk(ROOT):
        dirs[:] = [d for d in dirs if d not in EXCLUDE_DIRS and not d.startswith(".")]
        for fname in files:
            ext = fname.rsplit(".", 1)[-1] if "." in fname else ""
            if ext not in counters:
                continue
            filepath = os.path.join(root, fname)
            try:
                code, comment, blank = counters[ext](filepath)
            except Exception:
                continue
            totals["code"] += code
            totals["comment"] += comment
            totals["blank"] += blank
            totals["files"] += 1

            if ext not in by_ext:
                by_ext[ext] = {"code": 0, "comment": 0, "blank": 0, "files": 0}
            by_ext[ext]["code"] += code
            by_ext[ext]["comment"] += comment
            by_ext[ext]["blank"] += blank
            by_ext[ext]["files"] += 1

    total_all = totals["code"] + totals["comment"] + totals["blank"]

    print("=" * 60)
    print("项目代码统计")
    print("=" * 60)
    print(f"{'类型':<10} {'文件数':>6} {'代码行':>8} {'注释行':>8} {'空白行':>8} {'总行数':>8}")
    print("-" * 60)

    for ext in sorted(by_ext.keys()):
        s = by_ext[ext]
        t = s["code"] + s["comment"] + s["blank"]
        print(f"{ext:<10} {s['files']:>6} {s['code']:>8} {s['comment']:>8} {s['blank']:>8} {t:>8}")

    print("-" * 60)
    print(f"{'合计':<10} {totals['files']:>6} {totals['code']:>8} {totals['comment']:>8} {totals['blank']:>8} {total_all:>8}")
    print("=" * 60)

if __name__ == "__main__":
    main()

File: scripts/count_lines.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

var excludeDirs = map[string]bool{
	".git": true, "vendor": true, "node_modules": true,
	"target": true, "__pycache__": true, ".zenmind": true,
}

var counters = map[string]func(string) (code, comment, blank int){
	".go":  countCStyle,
	".rs":  countCStyle,
	".yml": countHashStyle,
	".md":  countMarkdown,
}

func countCStyle(path string) (code, comment, blank int) {
	f, err := os.Open(path)
	if err != nil {
		return 0, 0, 0
	}
	defer f.Close()

	inBlock := false
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			blank++
			continue
		}
		if inBlock {
			comment++
			if strings.Contains(line, "*/") {
				inBlock = false
			}
			continue
		}
		if strings.HasPrefix(line, "//") {
			comment++
			continue
		}
		if strings.HasPrefix(line, "/*") || strings.HasPrefix(line, "//*") {
			comment++
			if !strings.Contains(line, "*/") {
				inBlock = true
			}
			continue
		}
		code++
	}
	return
}

func countHashStyle(path string) (code, comment, blank int) {
	f, err := os.Open(path)
	if err != nil {
		return 0, 0, 0
	}
	defer f.Close()

	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			blank++
		} else if strings.HasPrefix(line, "#") {
			comment++
		} else {
			code++
		}
	}
	return
}

func countMarkdown(path string) (code, comment, blank int) {
	f, err := os.Open(path)
	if err != nil {
		return 0, 0, 0
	}
	defer f.Close()

	sc := bufio.NewScanner(f)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			blank++
		} else {
			comment++ // markdown 全算文档/注释
		}
	}
	return
}

func main() {
	root := "."
	if len(os.Args) > 1 {
		root = os.Args[1]
	}

	type stat struct {
		files, code, comment, blank int
	}
	total := stat{}
	byExt := map[string]*stat{}

	_ = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			name := info.Name()
			if excludeDirs[name] || (len(name) > 0 && name[0] == '.') {
				return filepath.SkipDir
			}
			return nil
		}
		ext := filepath.Ext(info.Name())
		fn, ok := counters[ext]
		if !ok {
			return nil
		}
		code, comment, blank := fn(path)
		total.files++
		total.code += code
		total.comment += comment
		total.blank += blank

		if byExt[ext] == nil {
			byExt[ext] = &stat{}
		}
		s := byExt[ext]
		s.files++
		s.code += code
		s.comment += comment
		s.blank += blank
		return nil
	})

	fmt.Println(strings.Repeat("=", 64))
	fmt.Println("项目代码统计")
	fmt.Println(strings.Repeat("=", 64))
	fmt.Printf("%-8s %6s %8s %8s %8s %8s\n", "类型", "文件数", "代码行", "注释行", "空白行", "总行数")
	fmt.Println(strings.Repeat("-", 64))

	for _, ext := range []string{".go", ".rs", ".yml", ".md"} {
		s := byExt[ext]
		if s == nil {
			continue
		}
		t := s.code + s.comment + s.blank
		fmt.Printf("%-8s %6d %8d %8d %8d %8d\n", ext, s.files, s.code, s.comment, s.blank, t)
	}

	all := total.code + total.comment + total.blank
	fmt.Println(strings.Repeat("-", 64))
	fmt.Printf("%-8s %6d %8d %8d %8d %8d\n", "合计", total.files, total.code, total.comment, total.blank, all)
	fmt.Println(strings.Repeat("=", 64))

	srcFiles := total.files
	if s, ok := byExt[".md"]; ok {
		srcFiles -= s.files
	}
	srcCode := total.code
	srcComment := total.comment
	srcBlank := total.blank
	if s, ok := byExt[".md"]; ok {
		srcCode -= s.code
		srcComment -= s.comment
		srcBlank -= s.blank
	}
	srcAll := srcCode + srcComment + srcBlank
	fmt.Println()
	fmt.Println("源代码文件（.go + .rs + .yml，不含文档）:")
	fmt.Printf("  文件数: %d, 代码行: %d, 注释行: %d, 空白行: %d, 总行数: %d\n",
		srcFiles, srcCode, srcComment, srcBlank, srcAll)
}
